aten/src/ATen/core/LazyTensor.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <ATen/core/TensorBody.h>
+#include <ATen/core/ivalue.h>
+#include <ATen/core/function_schema.h>
+#include <c10/core/TensorImpl.h>
+
+namespace at {
+
+struct CAFFE2_API LazyTensorImpl : public TensorImpl {
+  explicit LazyTensorImpl(TensorTypeSet ts)
+      : TensorImpl(
+            ts.add(TensorTypeId::LazyTensorId),
+            caffe2::TypeMeta::Make<float>(), // Hack to support gradients
+            c10::Device(c10::DeviceType::CPU)) {}
+
+  explicit LazyTensorImpl(at::Tensor const& inp)
+      : LazyTensorImpl(inp.type_set()) {
+    inps_.emplace_back(inp);
+  }
+
+  explicit LazyTensorImpl(
+      c10::optional<c10::FunctionSchema> schema,
+      std::vector<c10::IValue> const& inps,
+      size_t index = 0)
+      : LazyTensorImpl([&inps]() {
+          TensorTypeSet ts(TensorTypeId::LazyTensorId);
+          for (const auto& inp : inps) {
+            if (inp.isTensor()) {
+              ts = ts | inp.toTensor().type_set();
+            }
+          }
+          return ts;
+        }()) {
```

Comment:
Here's a hot take: LazyTensor's TensorTypeSet should ONLY ever be LazyTensorId. None of the tensor ids of its inputs. Why? Because we never redispatch on a lazy tensor, we only ever do operations on the tensors it holds. Type set is a sort of "is a" relationship, not a "has a" relationship.

Reply:
It's nice to know the underlying (lower precedent) TensorTypeSet that is being made lazy -- e.g. if its a GlowTensorId or CPUTensorId

Reply:
Yeah, this is an occurrence of what I call "dispatch key" versus "type" confusion. If TensorTypeSet was called DispatchKeySet would you still be so eager to put CPUDispatchKey in it? I think not!

Reply:
I still believe in this comment. I don't think the lower TensorTypeSet should be in the lazy TensorTypeSet.